File: 7. Array/054.py
from typing import List

class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        if not matrix:
            return []
        position_i = position_j = 0
        d = 0 # diretion flag
        up_bound = -1
        left_bound = -1
        down_bound = len(matrix)
        right_bound = len(matrix[0])
        res = [0] * (down_bound*right_bound)
        index = 0
        while down_bound-up_bound>1 and right_bound-left_bound>1:
            res[index] = matrix[position_i][position_j]
            index += 1
            if d == 0: # go right now
                if position_j+1 == right_bound:
                    d = 1
                    position_i += 1
                    up_bound += 1
                else:
                    position_j += 1
            elif d == 1: # go down now
                if position_i+1 == down_bound:
                    d = 2
                    position_j -= 1
                    right_bound -= 1
                else:
                    position_i += 1
            elif d == 2: # go left now
                if position_j-1 == left_bound:
                    d = 3
                    position_i -= 1
                    down_bound -= 1
                else:
                    position_j -= 1
            else: # go up now
                if position_i-1 == up_bound:
                    d = 0
                    position_j += 1
                    left_bound += 1
                else:
                    position_i -= 1
        return res

if __name__ == "__main__":
    a = [[1],[2],[3]]
    print(Solution().spiralOrder(a))

File: C++/054.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> res;
        if(matrix.size() == 0 or matrix.at(0).size() == 0){
            return res;
        }

        int left_bound{0}, right_bound{matrix.at(0).size()-1};
        int up_bound{0}, down_bound{matrix.size()-1};
        int r{0};
        int c{0};

        while(true){
            // move right
            while(c < right_bound){
                res.push_back(matrix.at(r).at(c));
                ++c;
            }
            res.push_back(matrix.at(r).at(c));
            ++r;
            ++up_bound;
            if(left_bound > right_bound or up_bound > down_bound)
                break;

            // move down
            while(r < down_bound){
                res.push_back(matrix.at(r).at(c));
                ++r;
            }
            res.push_back(matrix.at(r).at(c));
            --c;
            --right_bound;
            if(left_bound > right_bound or up_bound > down_bound)
                break;

            // move left
            while(c > left_bound){
                res.push_back(matrix.at(r).at(c));
                --c;
            }
            res.push_back(matrix.at(r).at(c));
            --r;
            --down_bound;
            if(left_bound > right_bound or up_bound > down_bound)
                break;

            // move up
            while(r > up_bound){
                res.push_back(matrix.at(r).at(c));
                --r;
            }
            res.push_back(matrix.at(r).at(c));
            ++c;
            ++left_bound;
            if(left_bound > right_bound or up_bound > down_bound)
                break;
        }
        return res;
    }
};

int main() {
    vector<vector<int>> matrix{{1,2,3,4},{5,6,7,8},{9,10,11,12}};
    Solution s;
    vector<int> res = s.spiralOrder(matrix);
    for(const auto &i:res){
        cout<<i<<" ";
    }
    cout<<endl;
    return 0;
}